### PR TITLE
update boost digest type

### LINKED ISSUE #1993: MacOS Fails to build on boost@1.86
I started running into this issue today when building esbmc.

> `src/util/crypto_hash.cpp:47:26: error: non-const lvalue reference to type 'digest_type' (aka 'unsigned char[20]') cannot bind to a value of unrelated type 'unsigned int[5]'
>   p_crypto->s.get_digest(hash);

> /opt/homebrew/include/boost/uuid/detail/sha1.hpp:179:43: note: passing argument to parameter 'digest' here
> inline void sha1::get_digest(digest_type& digest)`

On investigating, it seems by boost version was upgraded from 1.85 to 1.86.
The change log has this particular line of code from 1.85 to 1.86
`typedef unsigned int(digest_type)[5]; => typedef unsigned char digest_type[ 20 ];`

reverting to boost 1.85 works.

### PR DESCRIPTION
**Pull request by XLiZHI**
```
  size_t to_size_t() const
  {
    size_t result = hash[0];
    for (int i = 1; i < 20; i++)
      // Do we care about overlaps?
      result ^= hash[i];
    return result;
  }
```
I'm not sure about this. Unsigned int: 32; Unsigned char: 8. Does this method need to be changed?

### CODE REVIEW

**Comment by rafaelsamenezes** [comment]
Could you give a bit more context? Why are you increasing the size to 20?

Edit: just saw the issue.

File: src/util/crypto_hash.h [hunk 1]
```diff
@@ -10,10 +10,24 @@ class crypto_hash
 {
 public:
   std::shared_ptr<crypto_hash_private> p_crypto;
+#if BOOST_VERSION >= 108600
```
Comment (rafaelsamenezes) [inline src/util/crypto_hash.h]: Could you just use a define or a static constant for the array length? 

This way you dont need to repeat this ifdef.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

File: src/util/crypto_hash.h (hunk 1 above)
Comment (XLiZHI) [reply src/util/crypto_hash.h]: done

**Comment by lucasccordeiro** [comment]
@XLiZHI: Does this PR fixes this issue https://github.com/esbmc/esbmc/issues/1993?

**Comment by XLiZHI** [comment]
> @XLiZHI: Does this PR fixes this issue #1993?

Yes

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @XLiZHI.

**Pull request merged**